File: backend/feature/auth/src/main/java/com/veda/server/service/AccountService.java
package com.veda.server.service;

import com.veda.server.HashEncoder;
import com.veda.server.JwtService;
import com.veda.server.dto.request.CheckEmailRequest;
import com.veda.server.dto.request.RegisterRequest;
import com.veda.server.dto.response.AuthResponse;
import com.veda.server.dto.response.AvailabilityResponse;
import com.veda.server.exception.UserAlreadyExists;
import com.veda.server.model.User;
import com.veda.server.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;

@Service
@RequiredArgsConstructor
public class AccountService {

    private final UserRepository userRepository;
    private final HashEncoder encoder;
    private final JwtService jwtService;
    private final TokenService tokenService;

    @Transactional
    public AuthResponse createAccount(RegisterRequest request) {
        if (userRepository.existsByEmail(request.email())) {
            throw new UserAlreadyExists("User already exists");
        }

        User user = new User();
        user.setEmail(request.email());
        user.setPassword(encoder.encode(request.password()));
        user.setCratedAt(Instant.now());
        user.setUpdatedAt(Instant.now());

        User savedUser = userRepository.save(user);

        String accessToken = jwtService.generateAccessToken(savedUser.getId().toString());
        String refreshToken = jwtService.generateRefreshToken(savedUser.getId().toString());

        tokenService.saveRefreshToken(savedUser, refreshToken, jwtService.getRefreshTokenValidityMs());

        return new AuthResponse(refreshToken, accessToken);
    }

    @Transactional(readOnly = true)
    public AvailabilityResponse checkEmail(CheckEmailRequest request) {
        boolean exists = userRepository.existsByEmail(request.email());
        return new AvailabilityResponse(!exists, exists ? "Email already exists" : null);
    }
}

File: src/main/kotlin/com/titsuko/server/service/AccountService.kt
package com.titsuko.server.service

import com.titsuko.server.dto.request.CheckEmailRequest
import com.titsuko.server.dto.request.RegisterRequest
import com.titsuko.server.dto.response.AccountResponse
import com.titsuko.server.dto.response.AuthResponse
import com.titsuko.server.dto.response.AvailabilityResponse
import com.titsuko.server.exception.InvalidTokenException
import com.titsuko.server.exception.UserAlreadyExistsException
import com.titsuko.server.exception.UserNotFoundException
import com.titsuko.server.model.Profile
import com.titsuko.server.model.User
import com.titsuko.server.repository.ProfileRepository
import com.titsuko.server.repository.UserRepository
import com.titsuko.server.security.HashEncoder
import com.titsuko.server.security.JwtService
import com.titsuko.server.security.RefreshService
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class AccountService(
    private val userRepository: UserRepository,
    private val profileRepository: ProfileRepository,
    private val jwtService: JwtService,
    private val refreshService: RefreshService,
    private val hashEncoder: HashEncoder
) {

    @Transactional
    fun register(request: RegisterRequest): AuthResponse {
        if (userRepository.findByEmail(requireNotNull(request.email)) != null) {
            throw UserAlreadyExistsException()
        }

        val hashedPassword = hashEncoder.encode(requireNotNull(request.password)).toString()
        val (firstName, lastName) = parseFullName(requireNotNull(request.fullName))

        val profile = profileRepository.save(Profile(
            firstName = firstName,
            lastName = lastName
        ))

        val user = userRepository.save(User(
            email = request.email,
            password = hashedPassword,
            profile = profile
        ))

        val accessToken = jwtService.generateAccessToken(user.email)
        val refreshToken = refreshService.createToken(user)

        return AuthResponse(
            accessToken = accessToken.first,
            refreshToken = refreshToken,
            expiresIn = accessToken.second
        )
    }

    private fun parseFullName(fullName: String): Pair<String, String> {
        val parts = fullName.trim().split("\\s+".toRegex(), 2)
        return parts[0] to (parts.getOrNull(1) ?: "")
    }

    @Transactional(readOnly = true)
    fun getProfile(): AccountResponse {
        val authentication = SecurityContextHolder.getContext().authentication
        val email = authentication?.name
            ?: throw InvalidTokenException("Session expired or invalid")

        val user = userRepository.findByEmail(email)
            ?: throw UserNotFoundException(email)

        return AccountResponse(
            firstName = user.profile.firstName,
            lastName = user.profile.lastName,
            createdAt = user.createdAt
        )
    }

    @Transactional(readOnly = true)
    fun checkEmail(request: CheckEmailRequest): AvailabilityResponse {
        val exists = userRepository.findByEmail(requireNotNull(request.email)) != null
        return AvailabilityResponse(
            available = !exists,
            message = if (exists) "Email already exists" else "Email is available"
        )
    }
}
